File: backend/app/routers/tasks.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.models import Task
from app.schemas.projects.task import TaskCreate, TaskUpdate, TaskOut, TaskWithDetails

router = APIRouter(prefix="/tasks", tags=["Tasks"])


@router.get("/", response_model=list[TaskOut])
def get_all_tasks(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return db.query(Task).offset(skip).limit(limit).all()


@router.get("/{task_id}", response_model=TaskWithDetails)
def get_task(task_id: int, db: Session = Depends(get_db)):
    task = db.query(Task).filter(Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.get("/project/{project_id}", response_model=list[TaskWithDetails])
def get_by_project(project_id: int, db: Session = Depends(get_db)):
    return db.query(Task).filter(Task.project_id == project_id).all()


@router.get("/assigned/{user_id}", response_model=list[TaskOut])
def get_by_assignee(user_id: int, db: Session = Depends(get_db)):
    return db.query(Task).filter(Task.assigned_to == user_id).all()


@router.post("/", response_model=TaskOut, status_code=201)
def create_task(data: TaskCreate, db: Session = Depends(get_db)):
    task = Task(**data.model_dump())
    db.add(task)
    db.commit()
    db.refresh(task)
    return task


@router.patch("/{task_id}", response_model=TaskOut)
def update_task(task_id: int, data: TaskUpdate, db: Session = Depends(get_db)):
    task = db.query(Task).filter(Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(task, field, value)
    db.commit()
    db.refresh(task)
    return task


@router.delete("/{task_id}", status_code=204)
def delete_task(task_id: int, db: Session = Depends(get_db)):
    task = db.query(Task).filter(Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    db.delete(task)
    db.commit()

File: frontend/lib/api/tasks.ts
/**
 * lib/api/tasks.ts
 */

import { apiRequest } from "./client";

export interface TaskOut {
  id: number;
  title?: string;
  description?: string;
  project_id?: number;
  status_id?: number;
  assigned_to?: number;
  created_at?: string;
}

export interface TaskWithDetails extends TaskOut {
  status?: { id: number; name: string } | null;
  assignee?: { id: number; name?: string; username?: string } | null;
}

export interface TaskCreate {
  title: string;
  project_id: number;
  description?: string;
  status_id?: number;
  assigned_to?: number;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  project_id?: number;
  status_id?: number;
  assigned_to?: number;
}

export async function getTasks(skip = 0, limit = 100): Promise<TaskOut[]> {
  return apiRequest<TaskOut[]>("/tasks", { params: { skip, limit } });
}

export async function getTask(taskId: number): Promise<TaskWithDetails> {
  return apiRequest<TaskWithDetails>(`/tasks/${taskId}`);
}

export async function getTasksByProject(projectId: number): Promise<TaskWithDetails[]> {
  return apiRequest<TaskWithDetails[]>(`/tasks/project/${projectId}`);
}

export async function getTasksByAssignee(userId: number): Promise<TaskOut[]> {
  return apiRequest<TaskOut[]>(`/tasks/assigned/${userId}`);
}

export async function createTask(data: TaskCreate): Promise<TaskOut> {
  return apiRequest<TaskOut>("/tasks", { method: "POST", body: data });
}

export async function updateTask(taskId: number, data: TaskUpdate): Promise<TaskOut> {
  return apiRequest<TaskOut>(`/tasks/${taskId}`, { method: "PATCH", body: data });
}

export async function deleteTask(taskId: number): Promise<void> {
  return apiRequest<void>(`/tasks/${taskId}`, { method: "DELETE" });
}
